### ISSUE TITLE
[BUG] Optimizer single item not returned

### ISSUE THREAD

**Issue by keithchew** [RediSearch/RediSearch#4428]
Testing on v2.8.11, I found an edge case where WITHOUTCOUNT and WITHCOUNT return different results.

WITHOUTCOUNT:
```
> ft.profile testidx search query "@id:[153977 153977] @user:{testuser1}" sortby timestamp asc WITHOUTCOUNT return 1 user limit 0 1

1) 1) "0"
2) 1) 1) "Total profile time"
      2) "0"
   2) 1) "Parsing time"
      2) "0"
   3) 1) "Pipeline creation time"
      2) "0"
   4) 1) "Warning"
   5) 1) "Iterators profile"
      2) 1) "Type"
         2) "OPTIMIZER"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "0"
         7) "Optimizer mode"
         8) "Query partial range"
         9) "Child iterator"
         10) 1) "Type"
            2) "INTERSECT"
            3) "Time"
            4) "0"
            5) "Counter"
            6) "1"
            7) "Child iterators"
            8) 1) "Type"
               2) "TAG"
               3) "Term"
               4) "testuser1"
               5) "Time"
               6) "0"
               7) "Counter"
               8) "3"
               9) "Size"
               10) "814"
            9) 1) "Type"
               2) "NUMERIC"
               3) "Term"
               4) "144877 - 153979"
               5) "Time"
               6) "0"
               7) "Counter"
               8) "1"
               9) "Size"
               10) "2081"
   6) 1) "Result processors profile"
      2) 1) "Type"
         2) "Index"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "0"
      3) 1) "Type"
         2) "Sorter"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "0"
      4) 1) "Type"
         2) "Loader"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "0"
```

WITHCOUNT:
```
> ft.profile testidx search query "@id:[153977 153977] @user:{testuser1}" sortby timestamp asc WITHCOUNT return 1 user limit 0 1

1) 1) "1"
   2) "namespace:testuser1"
   3) 1) "user"
      2) "testuser1"
2) 1) 1) "Total profile time"
      2) "0"
   2) 1) "Parsing time"
      2) "0"
   3) 1) "Pipeline creation time"
      2) "0"
   4) 1) "Warning"
   5) 1) "Iterators profile"
      2) 1) "Type"
         2) "INTERSECT"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "1"
         7) "Child iterators"
         8) 1) "Type"
            2) "TAG"
            3) "Term"
            4) "testuser1"
            5) "Time"
            6) "0"
            7) "Counter"
            8) "3"
            9) "Size"
            10) "814"
         9) 1) "Type"
            2) "NUMERIC"
            3) "Term"
            4) "144877 - 153979"
            5) "Time"
            6) "0"
            7) "Counter"
            8) "1"
            9) "Size"
            10) "2081"
   6) 1) "Result processors profile"
      2) 1) "Type"
         2) "Index"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "1"
      3) 1) "Type"
         2) "Sorter"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "1"
      4) 1) "Type"
         2) "Loader"
         3) "Time"
         4) "0"
         5) "Counter"
         6) "1"
```

The dataset contains 80000+ records, and the criteria above only matches 1 record. I have not been able to create a smaller test dataset to reproduce this case, but I have tracked it down with some logging:
```
10:M 08 Feb 2024 09:46:46.662 # <module> rc2 [2] numericRes->docId [63892] SkipTo [63873] nf->limit [868] init
10:M 08 Feb 2024 09:46:46.662 # <module> childRes->docId [63873] numericRes->docId [63892]
10:M 08 Feb 2024 09:46:46.662 # <module> rc1 [0] SkipTo [63892]
10:M 08 Feb 2024 09:46:46.662 # <module> rc1 [0], break...
10:M 08 Feb 2024 09:46:46.662 # <module> rc1 [0], break...
10:M 08 Feb 2024 09:46:46.662 # <module> rc1 [0], break...
10:M 08 Feb 2024 09:46:46.662 # <module> rc1 [0], break...
10:M 08 Feb 2024 09:46:46.662 # <module> it->numIterations [9] heap_count(it->heap) [0]
```

It looks like when there is only 1 record in the INTERSECT, the rewind did not work, as subsequent reads (after rewind) just returns INDEXREAD_EOF straight away. If I modify the above query to be @id:[153977 153978], which still only matches 1 record, both WITH and WITHOUT returns the correct 1 record.

Very odd case indeed!

**Comment by GuyAv46** [comment]
Very nice. I asked you to add a test that will fail without the fix. A good place will be in `test_issues.py`, naming it `test_issue_4428`.
If Python is not your strong side you can also describe a toy example here and I’ll write the test and add it to the PR

**Comment by GuyAv46** [comment]
I approved the PR, it is now merged to master and 2.8.
I added an internal task to add some test that causes the IR iterator to rewind, but as this fix is obvious, it can be merged on its own.
Thanks for the contribution!
